src/protocol/sort/secureapplyinv.rs

```diff
@@ -0,0 +1,171 @@
+    shuffle::{generate_random_permutation, Shuffle},
+};
+use tokio::try_join;
+
+/// This is APPLYINV(Algorithm 4) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This is a protocol that applies the inverse of a shared permutation to a shared-vector representation.
+/// Input: Each helpers know their own secret shares of input and permutation
+/// Output: At the end of the protocol, all helpers receive inputs after the permutation is applied
+#[derive(Debug)]
+#[embed_doc_image("secureapplyinv", "images/sort/secureapplyinv.png")]
+pub struct SecureApplyInv<'a, F> {
+    input: &'a mut Vec<Replicated<F>>,
+    permutation: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> SecureApplyInv<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, permutation: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { input, permutation }
+    }
+    /// This algorithm applies a permutation on the inputs. The permutation is not available in clear and
+    /// we do not want to reveal the actual permutation to any of the helpers while applying it.
+    /// Steps
+    /// ![Secure Apply Inv steps][secureapplyinv]
+    /// 1. Generate a permutation
+    /// 2. Secret shared permutation is shuffled with a permutation
+    /// 3. Secret shared value is shuffled using the same permutation as 1
+    /// 4. The permutation is revealed
+    /// 5. All helpers call `apply` to apply the permutation locally.
+    #[allow(dead_code)]
+    pub async fn execute<N: Network>(
+        &mut self,
+        ctx: &ProtocolContext<'_, N>,
+    ) -> Result<(), BoxError> {
+        // 1. Generate permutation
+        let permutations = generate_random_permutation(self.input.len(), &ctx.prss());
+        // 2. Shuffle permutations
+        let mut shuffle_permutation = Shuffle::new(self.permutation);
+        let h0_future = shuffle_permutation.execute(ctx.narrow(&ShufflePermutation), &permutations);
+
+        // 3. Shuffle inputs
+        let mut shuffle_input = Shuffle::new(self.input);
+        let h1_future = shuffle_input.execute(ctx.narrow(&ShuffleInputs), &permutations);
+
+        try_join!(h0_future, h1_future).unwrap();
+
+        // 4. Reveal permutation to all helpers
+        let permutation = self
+            .reveal_permutations(&ctx.narrow(&RevealPermutation))
+            .await
+            .unwrap();
+        let mut perms = Vec::new();
+        // TODO(richa) : This is ugly, converted F -> u128 -> usize needed for getting an appropriate Permutation to apply.
+        // Besides this makes the code dependent on Permutation struct needs which is not a good design choice.
+        for i in &permutation {
+            perms.push(i.as_u128() as usize);
+        }
+        let mut permutation = Permutation::oneline(perms);
+
+        // 5. apply the permutation on the shuffled input
+        // The paper suggests taking apply_inv. However, Permutation::oneline returns an inverted permutation, so we are negating
```

Comment:
is it worth deviating from the paper here? would `Permutation::oneline().inverse()` do what you want?

Reply:
I'm totally confused by this comment.

Reply:
Removing this comment.